### PR TITLE
Add Charmstore v5 and multi-series support

### PR DESCRIPTION
**Pull request by hatched**
Switches to use the Charmstore v5 endpoint and adds multi-series support to the UI for Juju 1 and 2.

When deploying a multi-series charm the GUI will pick, by default, the first series in the charm series list which is considered to be the preferred one set by the author. If the user wishes to change this series they can open the Configuration pane in the inspector and modify the dropdown. 

When manually placing units for multi-series charms the machine unit will be locked to whatever the series of the multi-series charm is set to when dropping the unit on the machine. This means that if you change the series in the inspector after the fact we unplace that applications units, switch to the machine view, and notify the user of this action. 

Using multi-series charms is transparent to the user and the GUI maintains the easy one click 'add-to-canvas' functionality so that the user doesn't ever have to consider the series if they are not explicitly interested in doing so.

### CODE REVIEW

**Comment by hatched** [comment]
#### To QA

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5440/
Test FAILed.

File: jujugui/static/gui/src/app/app.js [hunk 1]
```diff
@@ -1110,6 +1110,7 @@ YUI.add('juju-gui', function(Y) {
         var charm = app.db.charms.getById(service.get('charm'));
         var relatableApplications = utils.getRelatableApplications(
           this.db, models.getEndpoints(service, this.endpointsController));
+        const ecs = this.env.get('ecs');
```
Comment (hatched) [inline jujugui/static/gui/src/app/app.js]: I started using es6 variable definitions part way through this work. We should work towards using these so as more branches go on we'll update existing code.

File: jujugui/static/gui/src/app/assets/javascripts/jujulib/charmstore.js [hunk 2]
```diff
@@ -160,15 +160,15 @@ var module = module;
           charmMeta = meta['charm-metadata'],
           charmConfig = meta['charm-config'],
           bundleMeta = meta['bundle-metadata'],
-          bzrOwner = extraInfo['bzr-owner'];
+          owner = meta.owner && meta.owner.User;
```
Comment (hatched) [inline jujugui/static/gui/src/app/assets/javascripts/jujulib/charmstore.js]: Owner was updated as a side effect of this branch which should fix the issues we were seeing where promulgated charms for certain users wouldn't properly link correctly.

File: jujugui/static/gui/src/app/components/search-results-item/search-results-item.js [hunk 3]
```diff
@@ -88,7 +88,7 @@ YUI.add('search-results-item', function(Y) {
               className="list-icons__image"
               alt={service.displayName} />
             <span className="tooltip__tooltip">
-              <span className="tooltip__inner">
+              <span className="tooltip__inner tooltip__inner--down">
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/search-results-item/search-results-item.js]: Drive-by to fix a tooltip styling bug.

File: jujugui/static/gui/src/app/utils/changes-utils.js [hunk 4]
```diff
@@ -117,7 +117,7 @@ YUI.add('changes-utils', function(Y) {
           changeItem.description = ' ' + ghostService.get('name') +
               ' has been added.';
           break;
-        case '_destroyService':
+        case '_destroyApplication':
```
Comment (hatched) [inline jujugui/static/gui/src/app/utils/changes-utils.js]: Drive-by

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5447/
Test FAILed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5449/
Test PASSed.

File: jujugui/static/gui/src/test/test_bundle_importer.js [hunk 5]
```diff
@@ -18,7 +18,7 @@ with this program.  If not, see <http://www.gnu.org/licenses/>.
 
 'use strict';
 
-describe('Bundle Importer', function() {
+describe.skip('Bundle Importer', function() {
```
Comment (hatched) [inline jujugui/static/gui/src/test/test_bundle_importer.js]: See email about bundlechanges lib

File: jujugui/static/gui/src/app/app.js (hunk 1 above)
Comment (frankban) [inline jujugui/static/gui/src/app/app.js]: That's great, but are you sure the ecs object is a constant? Oh, I read now that "the const declaration creates a read-only reference to a value. It does not mean the value it holds is immutable, just that the variable identifier cannot be reassigned". Great, this will confuse me forever...

File: jujugui/static/gui/src/app/components/entity-header/entity-header.js [hunk 6]
```diff
@@ -258,11 +258,27 @@ YUI.add('entity-header', function() {
...
+      single-series charms.
+
+      @method_generateSeriesList
+    */
+    _generateSeriesList: function() {
+      var series = this.props.entityModel.get('series');
+      if (!series) {
+        return null;
+      }
+      if (Array.isArray(series)) {
```
Comment (frankban) [inline jujugui/static/gui/src/app/components/entity-header/entity-header.js]: To avoid repeating the node code twice, what about:
```javascript
if (!Array.isArray(series)) {
  series  = [series];
}
return series.map(series => 
  <li key={series} className="entity-header__series">{series}</li>);
```

File: jujugui/static/gui/src/app/components/inspector-config/inspector-config.js [hunk 7]
```diff
@@ -287,6 +289,72 @@ YUI.add('inspector-config', function() {
...
+          <select
+            className="inspector-config__select"
+            onChange={this._handleSeriesChange}
+            value={this.state.series}>
+            {series.map(name =>
+              <option key={name} value={name}>{name}</option>)}
+          </select>
+          <span className="inspector-config__series-select-description">
+            Choose the series to deploy. This cannot be
+            changed once it is deployed
```
Comment (frankban) [inline jujugui/static/gui/src/app/components/inspector-config/inspector-config.js]: once the application is deployed?

File: jujugui/static/gui/src/app/extensions/autodeploy-extension.js [hunk 8]
```diff
@@ -47,7 +47,10 @@ YUI.add('autodeploy-extension', function(Y) {
           env = this.get('env') || this.env,
           unplacedUnits = db.units.filterByMachine(null);
       unplacedUnits.forEach(function(unit) {
-        var machine = this._createMachine();
+        // Get the application name from the unit id then grab the series
+        // for that application.
+        var series = db.services.getById(unit.id.split('/')[0]).get('series');
```
Comment (frankban) [inline jujugui/static/gui/src/app/extensions/autodeploy-extension.js]: Doesn't a unit have a service property already?

File: jujugui/static/gui/src/app/models/models.js [hunk 9]
```diff
@@ -607,7 +607,17 @@ YUI.add('juju-models', function(Y) {
         @type {Object}
         @default undefined
       */
-      activePlan: {}
+      activePlan: {},
+
+      /**
+        In a multi-series charm the service of the service will not always
```
Comment (frankban) [inline jujugui/static/gui/src/app/models/models.js]: the service of the service?
Please make this comment less confusing: what's stored here, and why?

File: jujugui/static/gui/src/app/store/env/api.js [hunk 10]
```diff
@@ -1198,6 +1199,11 @@ YUI.add('juju-env-api', function(Y) {
         'num-units': numUnits
       };
 
+      if (series) {
+        // We only want to supply the series when it is a multi-series charm.
```
Comment (frankban) [inline jujugui/static/gui/src/app/store/env/api.js]: Why? If series are passed I think we want to always provide it, and your code actually does that.

File: jujugui/static/gui/src/app/utils/environment-change-set.js [hunk 11]
```diff
@@ -1365,18 +1357,75 @@ YUI.add('environment-change-set', function(Y) {
...
+    /**
+      Takes a service id and then handles unplacing all of the uncommitted
+      units for that service.
+
+      @method unplaceServiceUnits
+      @param {String} serviceId The service id to unplace units for.
+      @returns {Array} Any units that had been unplaced, or an empty array.
+    */
+    unplaceServiceUnits: function(serviceId) {
+      return this.get('db').units
```
Comment (frankban) [inline jujugui/static/gui/src/app/utils/environment-change-set.js]: Alternatively `this.get('db').services.getById(serviceId).get('units').map(this.unplaceUnit(unit))`?

File: jujugui/static/gui/src/app/utils/environment-change-set.js [hunk 12]
```diff
@@ -1365,18 +1357,75 @@ YUI.add('environment-change-set', function(Y) {
...
+        // We only want to unplace units which have the matching
+        // service id and which are placed on machines.
+        .filter(unit => unit.service === serviceId && unit.machine)
+        // Unplace any units which match the criteria.
+        .map(unit => this.unplaceUnit(unit));
+    },
+
+    /**
+      removed the placed unit from the machine it's placed on.
```
Comment (frankban) [inline jujugui/static/gui/src/app/utils/environment-change-set.js]: s/removed/Remove/

File: jujugui/static/gui/src/app/utils/environment-change-set.js [hunk 13]
```diff
@@ -1365,18 +1357,75 @@ YUI.add('environment-change-set', function(Y) {
...
+      const db = this.get('db');
+      // Update the revived model to trigger events.
+      const unitModel = db.units.revive(unit);
+      unit = this._unplaceUnit(unit, unitModel);
+      db.units.free(unitModel);
+      return unit;
+    },
+
+    /**
+      removed the placed unit from the machine it's placed on.
```
Comment (frankban) [inline jujugui/static/gui/src/app/utils/environment-change-set.js]: s/removed/Removes/

File: jujugui/static/gui/src/app/views/ghost-deployer-extension.js [hunk 14]
```diff
@@ -127,6 +134,36 @@ YUI.add('ghost-deployer-extension', function(Y) {
     },
 
     /**
+      If we're using Juju 1 then we need to deploy a charm Id which has
```
Comment (frankban) [inline jujugui/static/gui/src/app/views/ghost-deployer-extension.js]: This is true, but does not explain the goal of the function.

File: jujugui/static/gui/src/app/views/ghost-deployer-extension.js [hunk 15]
```diff
@@ -127,6 +134,36 @@ YUI.add('ghost-deployer-extension', function(Y) {
...
+      the series defined in the URL. This is not required for Juju 2 as it
+      supports multi-series charms.
+
+      @method _addSeriesToCharmId
+      @param {String} charmId The charm id.
+      @param {String} series The series of the service.
+    */
+    _addSeriesToCharmId: function(charmId, series) {
+      let charmIdParts = charmId.replace('cs:', '').split('/');
```
Comment (frankban) [inline jujugui/static/gui/src/app/views/ghost-deployer-extension.js]: I would be nice to have a charm/bundle URL parsing library, returning an object with series, name, revision etc attributes, able to be re-serialized as an URL string. We have been having this splitting logic forever everywhere in the GUI. Not in this branch of course.

File: jujugui/static/gui/src/test/test_bundle_importer.js (hunk 5 above)
Comment (frankban) [inline jujugui/static/gui/src/test/test_bundle_importer.js]: See my reply to that email :-)

File: jujugui/static/gui/src/test/test_ghost_deployer_extension.js [hunk 16]
```diff
@@ -150,11 +152,20 @@ describe('Ghost Deployer Extension', function() {
...
-    assert.isNull(args[6]); // Machine placement.
+    assert.strictEqual(args[1], 'trusty'); // service series.
+    assert.strictEqual(args[2], 'ghost-service-id'); // Service name.
+    assert.deepEqual(args[3], {}); // Config.
+    assert.strictEqual(args[5], 0); // Number of units.
+    assert.deepEqual(args[6], {}); // Constraints.
+    assert.isNull(args[7]); // Machine placement.
+  });
+
+  it('properly adds the series to Juju 1 aplications', function() {
```
Comment (frankban) [inline jujugui/static/gui/src/test/test_ghost_deployer_extension.js]: Can we also test adding series to user owned charms without series?

**Comment by frankban** [comment]
👍 this is great Jeff thanks. Only minor suggestions.
I'll QA this asap.

**Comment by frankban** [comment]
QA: first issue:
when adding ubuntu in uncommtted state, changing the series in the inspector and then saving does not work, the series is still xenial.

**Comment by frankban** [comment]
QA: second issue, while trying to addd a bundle to the canvas: combo?app/assets/javascripts/yui/queue-promote/queue-promote-min.js&app/assets/javascripts/yui/io-q…:28 Uncaught TypeError: Cannot read property 'method' of undefined
Perhaps this is expected given current inability to deploy bundles, but worth mentioning.

**Comment by frankban** [comment]
Except for the two issues above, deployment of single-series and multi-series charm seems to work (with the first listed series in the case of multi-series, given the bug preventing me from switching series in the uncommitted inspector). Hope others will QA asap so that we can confirm the issue and possibly find problems I missed. We are almost there Jeff, thanks a lot.

**Comment by frankban** [comment]
Ah, forgot to mention, my QA has been done with Juju2 master tip and GIJoe.

File: jujugui/static/gui/src/app/app.js (hunk 1 above)
Comment (hatched) [inline jujugui/static/gui/src/app/app.js]: @frankban yeah it's quite an unfortunate name :(

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5453/
Test PASSed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5454/
Test PASSed.

File: jujugui/static/gui/src/app/components/entity-header/entity-header.js (hunk 6 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/entity-header/entity-header.js]: Good idea, updated

File: jujugui/static/gui/src/app/components/inspector-config/inspector-config.js (hunk 7 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/inspector-config/inspector-config.js]: done

File: jujugui/static/gui/src/app/extensions/autodeploy-extension.js (hunk 8 above)
Comment (hatched) [inline jujugui/static/gui/src/app/extensions/autodeploy-extension.js]: It does but the service name and id are not the same thing at this point - they are only the same once deployed (yay legacy code)

File: jujugui/static/gui/src/app/models/models.js (hunk 9 above)
Comment (hatched) [inline jujugui/static/gui/src/app/models/models.js]: fixed

File: jujugui/static/gui/src/app/store/env/api.js (hunk 10 above)
Comment (hatched) [inline jujugui/static/gui/src/app/store/env/api.js]: When this was originally written I had a different architecture to get to this point. That's been fixed because it was bad :) I'll update this comment because now the only issue is that if series is defined but `undefined` (somehow) then the deploy doesn't complete successfully.

File: jujugui/static/gui/src/app/utils/environment-change-set.js (hunk 11 above)
Comment (hatched) [inline jujugui/static/gui/src/app/utils/environment-change-set.js]: I used this approach because it's the same model list that we use when placing the units.

File: jujugui/static/gui/src/app/utils/environment-change-set.js (hunk 12 above)
Comment (hatched) [inline jujugui/static/gui/src/app/utils/environment-change-set.js]: fixed

File: jujugui/static/gui/src/app/utils/environment-change-set.js (hunk 13 above)
Comment (hatched) [inline jujugui/static/gui/src/app/utils/environment-change-set.js]: fixed

File: jujugui/static/gui/src/app/views/ghost-deployer-extension.js (hunk 14 above)
Comment (hatched) [inline jujugui/static/gui/src/app/views/ghost-deployer-extension.js]: Oops, expanded the description

File: jujugui/static/gui/src/app/views/ghost-deployer-extension.js (hunk 15 above)
Comment (hatched) [inline jujugui/static/gui/src/app/views/ghost-deployer-extension.js]: oh yes, definitely! We also need to start putting the 'utils' into logical modules - it's very unwieldy

File: jujugui/static/gui/src/test/test_ghost_deployer_extension.js (hunk 16 above)
Comment (hatched) [inline jujugui/static/gui/src/test/test_ghost_deployer_extension.js]: done!

**Comment by hatched** [comment]
Thanks for the review and qa @frankban I've addressed your comments and fixed the qa issue around actually changing the series in the inspector.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5455/
Test PASSed.

**Comment by kadams54** [comment]
When I deploy ubuntu, I see three items in the summary. One, "1 ubuntu unit has been added.", is about the unit being created, but the other two seem like duplicates, both about the charm/service/application: "cs:ubuntu-0 charm has been added." and "ubuntu has been added."

http://www.awesomescreenshot.com/image/1412553/74deccad737e334c01d61c5cf93fe126

Not sure if this is intentional.

Aside from this, QA is :ok:.

**Comment by hatched** [comment]
Thanks for the reviews and qa's! 

@kadams54 as discussed, there is no bug there it's just messaging in the deployment summary output, it should be a simple string change to update the messaging. I'll leave this for now as there is deployment summary work coming up.

:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge
